### PR TITLE
gh-125041: test_zlib: For s390x HW acceleration, only skip checking the compressed bytes

### PR DESCRIPTION
**Pull request by encukou**
* Issue: gh-125041

### CODE REVIEW

File: Misc/NEWS.d/next/Tests/2024-10-07-14-13-38.gh-issue-125041.PKLWDf.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Re-enable skipped tests for :mod:`zlib` on the s390x architecture: only skip
```
Comment (rruuaanng) [inline Misc/NEWS.d/next/Tests/2024-10-07-14-13-38.gh-issue-125041.PKLWDf.rst:1]: It seems that NEWS isn't required for unittest.

**Comment by encukou** [comment]
s390x buildbots were removed, so there's no way for me to test this right now :(
@vstinner, do you want to review this?

File: Misc/NEWS.d/next/Tests/2024-10-07-14-13-38.gh-issue-125041.PKLWDf.rst (hunk 1 above)
Comment (encukou) [reply Misc/NEWS.d/next/Tests/2024-10-07-14-13-38.gh-issue-125041.PKLWDf.rst:1]: Right, it's not required.

**Comment by encukou** [comment]
!buildbot s390x

File: Lib/test/test_zlib.py [hunk 2]
```diff
@@ -223,12 +227,14 @@ def test_keywords(self):
...
 
-    @skip_on_s390x
     def test_speech128(self):
         # compress more data
         data = HAMLET_SCENE * 128
         x = zlib.compress(data)
-        self.assertEqual(zlib.compress(bytearray(data)), x)
+        if not HW_ACCELERATED:
+            # With hardware acceleration, the compressed bytes
+            # might not be identical.
```
Comment (vstinner) [inline Lib/test/test_zlib.py]: nitpick: can you put the comment on the if instead? same comment below.

File: Lib/test/test_zlib.py [hunk 3]
```diff
@@ -33,8 +33,10 @@ def _zlib_runtime_version_tuple(zlib_version=zlib.ZLIB_RUNTIME_VERSION):
 ZLIB_RUNTIME_VERSION_TUPLE = _zlib_runtime_version_tuple()
 
 
-# bpo-46623: On s390x, when a hardware accelerator is used, using different
-# ways to compress data with zlib can produce different compressed data.
+
```
Comment (vstinner) [inline Lib/test/test_zlib.py]: nitpick: 3 empty lines might be too much, 2 is enough :-)

**Review by vstinner — APPROVED** [review]
LGTM. I just left minor coding style remarks.

I don't recall exactly the details, I skipped these tests 2 years ago. But from what I read, the change looks correct.

Sadly, I don't recall how to reproduce the issue: how to get access to a s390x machine where the hardware acceleration is enabled.

**Comment by encukou** [comment]
The new builders have HW acceleration -- the tests fail on `main` :)

File: Lib/test/test_zlib.py (hunk 2 above)
Comment (encukou) [inline Lib/test/test_zlib.py]:
```suggestion
        # With hardware acceleration, the compressed bytes
        # might not be identical.
        if not HW_ACCELERATED:
```

File: Lib/test/test_zlib.py [hunk 4]
```diff
@@ -286,7 +291,10 @@ def test_pair(self):
             x1 = co.compress(data)
             x2 = co.flush()
             self.assertRaises(zlib.error, co.flush) # second flush should not work
-            self.assertEqual(x1 + x2, datazip)
+            if not HW_ACCELERATED:
+                # With hardware acceleration, the compressed bytes might not
+                # be identical.
```
Comment (encukou) [inline Lib/test/test_zlib.py]:
```suggestion
            # With hardware acceleration, the compressed bytes might not
            # be identical.
            if not HW_ACCELERATED:
```

File: Lib/test/test_zlib.py (hunk 3 above)
Comment (encukou) [reply Lib/test/test_zlib.py]:
```suggestion
```

**Comment by encukou** [comment]
Sorry, not `main` -- I meant “the existing branch”: [3.10](https://buildbot.python.org/#/builders/1571/builds/10) in particular.

**Pull request merged**